### PR TITLE
Introduce zoom, bell and progress indicators to ATS

### PR DESCRIPTION
**Pull request by Don-Vito**
## PR Checklist
* [x] Closes https://github.com/microsoft/terminal/issues/8912
* [x] CLA signed. 
* [ ] Tests added/passed
* [ ] Documentation updated.
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already.

## Detailed Description of the Pull Request / Additional comments
* Introduced separate DataTemplate for rendering TabPaletteItem
* Introduced DataTemplateSelector assigning this template
* Introduced indicators as observable properties of the terminal Tab
* Bound TabPaletteItem to these properties
* Ensured that BindingUpdate is always called on CmdPal to avoid races

### CODE REVIEW

**Comment by Don-Vito** [comment]
IndicatorsInTabSwitch

**Comment by Don-Vito** [comment]
Hmm.. fails randomly with `Windows_UI_Xaml!CXcpDispatcher::Tick` throwing `E_NOINTERFACE` 🤔

**Comment by Don-Vito** [comment]
@carlos-zamora - do you have any idea why spell-check forgot programming basics:
image

**Comment by carlos-zamora** [comment]
> @carlos-zamora - do you have any idea why spell-check forgot programming basics:
> image

oh boy. That's weird... @zadjii-msft or @DHowett any thoughts?

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1088]: Idea: should we just define the setter for `IsPaneZoomed` to update the `_headerControl`? (and do the same for all these other ones).

File: src/cascadia/TerminalApp/PaletteItemTemplateSelector.h
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/PaletteItemTemplateSelector.h:2]: nit: mind adding an abstract here? Maybe mention the difference between the two templates and when to use one over the other.

**Comment by DHowett** [comment]
Sorry, you've got to merge main. :smile: The spellbot got updated, and some files got significantly moved around.

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.xaml:84]: Since we're now duplicating a good amount of styling, maybe define a style at the top of the file and use it on both? For this one in particular, you could create a style `BasedOn={ThemeResource KeyChordBorderStyle}`.

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.xaml:128]:
```suggestion
                            <!-- Spinner -->
                            <ColumnDefinition Width="Auto"/>
```
Is that what you mean?

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.xaml:136]: Can you make the `DataType` for the `DataTemplate` `local:TabPaletteItem` to get around all this casting? Or does that cause issues with the `PaletteItemTemplateSelector`?

**Review by carlos-zamora — APPROVED** [review]
Just had a few ideas to throw out there. Nothing crazy.

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/TerminalTab.cpp:1088]: It is possible.. though actually we need to do something else:
*  TabHeaderControl should expose DependencyProperties rather simple properties
* We need to bind these DependencyProperties to the tab properties (manually as no XAML is available).

Probably we should create a follow up task instead. WDYT?

Comment (carlos-zamora) [reply src/cascadia/TerminalApp/TerminalTab.cpp:1088]: _That_ sounds way better. I like that. Yeah, file a follow up task please.

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (Don-Vito) [reply src/cascadia/TerminalApp/CommandPalette.xaml]: 🤦 I meant gutter

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (Don-Vito) [reply src/cascadia/TerminalApp/CommandPalette.xaml:136]: I am not sure how to do it - as the data type the template gets is `FilteredCommand`, rather than `TabPaletteItem` (the cast is done on the property)

File: src/cascadia/TerminalApp/PaletteItemTemplateSelector.h
Comment (Don-Vito) [reply src/cascadia/TerminalApp/PaletteItemTemplateSelector.h:2]: Added method documentation instead

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (Don-Vito) [reply src/cascadia/TerminalApp/CommandPalette.xaml:84]: @carlos-zamora - can you please elaborate? :blush: I thought here I only use a style, while `BasedOn` is required when defining a new style. I am not strong at these things :blush:

File: src/cascadia/TerminalApp/TerminalTab.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/TerminalTab.cpp:1100]: The follow up task is here: https://github.com/microsoft/terminal/issues/9043

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/CommandPalette.xaml:84]: My bad. I thought both of the data templates were pretty similar. But there actually doesn't seem to be much overlap. If there was more overlap, I was hoping you could create a new style at the top, base it on `KeyChordBorderStyle`, and use the new style throughout both data templates.

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/CommandPalette.xaml:136]: _That feels so weird to me_. But I guess that's just how XAML works haha.

File: src/cascadia/TerminalApp/CommandPalette.xaml
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.xaml:428]: I've never seen this pattern before, but this is _brilliant_. TIL!

**Review by zadjii-msft — APPROVED** [review]
Okay I love this. The follow-up is a good idea. I don't care so much about blocking over the interface thing.

**Comment by zadjii-msft** [comment]
(I'll let Carlos have the last say before merging this, since you did specifically request he take another look)

**Review by carlos-zamora — APPROVED** [review]
Doubly approve! haha

**Review by DHowett — COMMENTED** [review]
Absolutely brilliant implementation.

**Pull request merged**
